internal/net/buffer.go

```diff
@@ -0,0 +1,222 @@
+// SPDX-FileCopyrightText: 2023 The Pion community <https://pion.ly>
+// SPDX-License-Identifier: MIT
+
+// Package net implements DTLS specific networking primitives.
+// NOTE: this package is an adaption of pion/transport/packetio. If possible,
+// the updates made in this repository will be reflected back upstream. If not,
+// it is likely that this will be moved to a public package in this repository.
+package net
+
+import (
+	"bytes"
+	"errors"
+	"io"
+	"net"
+	"sync"
+	"time"
+
+	"github.com/pion/transport/deadline"
+)
+
+// ErrTimeout indicates that deadline was reached before operation could be
+// completed.
+var ErrTimeout = errors.New("buffer: i/o timeout")
+
+// AddrPacket is a packet payload and the associated remote address from which
+// it was received.
+type AddrPacket struct {
+	addr net.Addr
+	data bytes.Buffer
+}
+
+// PacketBuffer is a circular buffer for network packets. Each slot in the
+// buffer supports a
+type PacketBuffer struct {
+	mutex sync.Mutex
+
+	packets     []AddrPacket
+	write, read int
+
+	// full indicates whether the buffer is full, which is needed to distinguish
+	// when the write pointer and read pointer are at the same index.
+	full bool
+
+	notify chan struct{}
+	closed bool
+
+	readDeadline *deadline.Deadline
+}
+
+// NewPacketBuffer creates a new PacketBuffer.
+func NewPacketBuffer() *PacketBuffer {
+	return &PacketBuffer{
+		readDeadline: deadline.New(),
+		// In the narrow context in which this package is currently used, there
+		// will always be at least one packet written to the buffer. Therefore,
+		// we opt to allocate with size of 1 during construction, rather than
+		// waiting until that first packet is written.
+		packets: make([]AddrPacket, 1),
+		full:    false,
+	}
+}
+
+// WriteTo writes a single packet to the buffer. The supplied address will
+// remain associated with the packet.
+func (b *PacketBuffer) WriteTo(p []byte, addr net.Addr) (int, error) {
+	b.mutex.Lock()
+
+	if b.closed {
+		b.mutex.Unlock()
+		return 0, io.ErrClosedPipe
+	}
+
+	var notify chan struct{}
+	if b.notify != nil {
+		notify = b.notify
+		b.notify = nil
+	}
+
+	// Check to see if we are full.
+	if b.full {
+		// If so, grow AddrPacket buffer.
+		var newPackets int
+		if len(b.packets) < 128 {
+			// Double the number of packets.
+			newPackets = len(b.packets)
+		} else {
+			// Increase the number of packets by 25%.
+			newPackets = len(b.packets) / 4
+		}
+		b.packets = append(b.packets[:b.write], append(make([]AddrPacket, newPackets), b.packets[b.write:]...)...)
```

Comment:
might it be more efficient and possibly easier to read to just allocate the total amount of new space and then `copy` over the existing packets followed by appending the new ones?

Reply:
sounds good. if anything it feels more readable